### ISSUE TITLE
Bioboxes should create a GUI for easy running bioboxes

### ISSUE THREAD

**Issue by michaelbarton** [bioboxes/rfc#94]
Creating a graphical user interface (GUI) for bioboxes would be the single
fastest thing to accelerate adoption. At present using Docker requires
knowledge of Docker and the command line. A GUI that someone can download and
then start their data through a biobox would put a huge amount of software in
the hands of many biologists.

This is not an immediate priority but I think is something we should eventually
consider.

**Comment by michaelbarton** [comment]
A GUI would also help editors/reviews to enforce the standard too.

**Comment by pditommaso** [comment]
Are you thinking to something like [Kitematic](https://kitematic.com/) ?

**Comment by pbelmann** [comment]
@pditommaso something like Kitematic would be great. It should just be a little bit adjusted to bioboxes. With features like:

* input fields for every interface variable (like fastq)
* choosing between bioboxes containers of one type (assembler, binning ....)
* or maybe even creating and saving your bioboxes pipeline by connecting containers in a gui
* with the new docker 1.5.0 features we could show resource usage statistics and compare them between containers of one type  

For a rich client application I would maybe consider  
* JavaFx with many nice [features](http://download.oracle.com/otndocs/products/javafx/2/samples/Ensemble/index.html) or
* [netbeans](https://netbeans.org/features/platform/features.html) for a fast development

Another option would be to build a browser based app that runs only on your local machine with the full power of web technologies like [bootstrap](http://getbootstrap.com/), [spring boot](http://projects.spring.io/spring-boot/), [angularjs](https://angularjs.org/) or [nodejs](https://nodejs.org/) .

**Comment by avilella** [comment]
@pditommaso it would be great if someone could have a look at http://kitematic.com . I don't use Mac, so I don't have the tools to try it out :-/

Some of the bullet points in @pbelmann comment could be covered by http://arvados.org , although arvados is a lot bigger in its feature set than the minimum we may want to try as a first off.

I have been playing with arvados, and maybe @tetron can give us a brief of how arvados could be used in this, but as a summary, arvados is the open-source stack that @tetron and others at curoverse use in a cloud setting. Taking the open source arvados and installing it locally would be a way of having a GUI for bioboxes apps.

Precisely along those lines, I have been looking into installing arvados locally on a Linux Laptop. It would be super-cool to have a demo ready for BOSC 2015 to show how, for each of the bioboxes we produce, we could execute them as an arvados "pipeline" using the Arvados GUI. I haven't successfully finished the installation, but @tetron and others are helping out with it:

https://arvados.org/issues/5461#note-5

I would be interested in knowing other people's opinion on this matter.

**Comment by michaelbarton** [comment]
> It would be super-cool to have a demo ready for BOSC 2015 to show how,
> for each of the bioboxes we produce

I think before BOSC 2015 we have to focus on getting some working bioboxes
and a website that provides documentation on how to create your own.

**Comment by jmchilton** [comment]
It might be worth also considering [Galaxy](http://galaxyproject.org/) as your GUI front-end - it is a great platform for this sort of thing. You could build Galaxy tool descriptions for [your boxes](https://github.com/apetkau/galaxy-hackathon-2014) and get many of the advantages you mentioned while also plugging into a very active bioinformatics focused developer community. You could plug multiple bioboxes into a pipeline/workflow built with a GUI with today's feature set. I would be happy to work with you on building deployment options that make this more accessible as well (perhaps recipes to build pre-configured Galaxy instances with certain bioboxes on Docker, Vagrant, and Cloud might be a way to make it accessible in many different ways).

**Comment by pditommaso** [comment]
@avilella I've tried Kitematic some months ago. It has a nice interface that shows the installed images and allows one to search for other images in the Docker hub. Also, it manages the installation of docker+boot2docker+virtualbox transparently, if it's not already installed. This could be a plus for novice users. Quite surely the Docker strategy is to provide an implementation of Kitematic for the Windows platform.

I'm less convinced on using Galaxy as presentation layer. In my opinion a GUI for Bioboxes should be independent from any specify pipeline/workflow platform.

**Comment by jjkoehorst** [comment]
Sorry for bumping up an old thread but has there been any interest to build an XML file for Galaxy? I personally would be a big fan and can look into the creation of a galaxy xml wrapper but first need to dive into the syntax (all possibilities) of the assemblers and validators for biobox.

**Comment by michaelbarton** [comment]
@jjkoehorst I think integration for galaxy would be a great idea. What would you need from us to get started?

**Comment by jjkoehorst** [comment]
I have been thinking of multiple ways, either via the biobox pip module which would require an "installation" or a a simple python script that creates the YAML files and mounts the right input files to the docker container (no folder mounting) see a draft example here: 

http://github.com/jjkoehorst/GalaxyBiobox

GalaxyBiobox (master) $ python yamlCreator.py --fastq myReadFolder/reads.fq.gz --type single --id test --output output_data/ --assembler bioboxes/idba --tmp tmp

What I think I need is wether this would be a wise way to go or via one of the already existing modules? 
Second what are all the YAML options, such that an XML can be created with this information.

**Comment by dridk** [comment]
I agree . I was thining a similiar tool to make pipeline runnable by biologist. 
I imagine a REST API server to run bioboxes with websockets to monitor pipelines. ( Bar progression, live log output). 
A pipeline can be described with a description.yaml file which contains, name, logo, description ... 
a Json file in Json-form specification, could describe how to build a form user interface. 

    field {
      id : "fastq_input"
     name: "fastq input"
     type : "file"
     description : "Put your fastq here "
    }

Here are some mokup ... If you are interesting.. we can build it ! 
![image](https://cloud.githubusercontent.com/assets/1911063/24557277/6984833a-1637-11e7-800a-ed97a1b1e98a.png)

![spectacle z26158](https://cloud.githubusercontent.com/assets/1911063/24557370/bf2a6bec-1637-11e7-9a25-ac5b9ac771c6.png)

**Comment by dridk** [comment]
I am an expert in Qt/C++ . That's mean I can build a nice user interface using Qml technolgy. That's mean : native application, cross platform and faster than html . 
And we only need a simple server with python tornado to expose a REST/websocker api.
For example, to run a pipeline : 

     POST http://api.bioboxes/velvet 
     BODY  { input : "/path/to/fastq/" , optionA : 324, optionB: 342} 

 BODY are generated from a formular which follow the Json-form specification. 

I imagine the client as a simple File browser . Select your pipeline, and from it, you can download/upload files.
